src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -0,0 +1,708 @@
+@"F(Expression<Func<int>>): () => 0
+F(Expression): () => String.Empty
+");
+        }
+
+        [Fact]
+        public void OverloadResolution_03()
+        {
+            var source =
+@"using System;
+using System.Linq.Expressions;
+class Program
+{
+    static void F(Expression e) { }
+    static void F(Expression<Func<int>> e) { }
+    static void Main()
+    {
+        F(delegate () { return 0; });
+        F(delegate () { return string.Empty; });
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            // PROTOTYPE: Should report CS1946: An anonymous method expression cannot be converted to an expression tree.
+            comp.VerifyDiagnostics(
+                // (9,11): error CS1945: An expression tree may not contain an anonymous method expression
+                //         F(delegate () { return 0; });
+                Diagnostic(ErrorCode.ERR_ExpressionTreeContainsAnonymousMethod, "delegate () { return 0; }").WithLocation(9, 11),
+                // (10,11): error CS1945: An expression tree may not contain an anonymous method expression
+                //         F(delegate () { return string.Empty; });
+                Diagnostic(ErrorCode.ERR_ExpressionTreeContainsAnonymousMethod, "delegate () { return string.Empty; }").WithLocation(10, 11));
+        }
+
+        [Fact]
+        public void NullableAnalysis_01()
+        {
+            var source =
+@"#nullable enable
+class Program
+{
+    static void Main()
+    {
+        System.Delegate d;
+        d = Main;
+        d = () => { };
+        d = delegate () { };
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NullableAnalysis_02()
+        {
+            var source =
+@"#nullable enable
+class Program
+{
+    static void Main()
+    {
+        object o;
+        o = (System.Delegate)Main;
+        o = (System.Delegate)(() => { });
+        o = (System.Delegate)(delegate () { });
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void TaskRunArgument()
+        {
+            var source =
+@"using System.Threading.Tasks;
+class Program
+{
+    static async Task F()
+    {	
+        await Task.Run(() => { });
+    }
+}";
+            var verifier = CompileAndVerify(source, parseOptions: TestOptions.RegularPreview);
+            var method = (MethodSymbol)verifier.TestData.GetMethodsByName()["Program.<>c.<F>b__0_0()"].Method;
+            Assert.Equal("void Program.<>c.<F>b__0_0()", method.ToTestDisplayString());
+            verifier.VerifyIL("Program.<>c.<F>b__0_0()",
+@"{
+  // Code size        1 (0x1)
+  .maxstack  0
+  IL_0000:  ret
+}");
+        }
+    }
+}
```

Comment:
Some test ideas:
1. the spec calls out behavior when the lambda or method group involves a modreq or modopt, but I couldn't find tests for that. ("`modopt()` or `modreq()` in the method group signature are ignored in the corresponding delegate type.")
2. make `Action`, `Func`, or `Delegate` missing
3. have we covered this section of the spec: "if [...] any of the parameter types or return are not valid type arguments"?
4. I'm not sure how to track this: we should remember to test delegate type inference once we support explicit return types. (I don't expect any problem though)
5. You already have tests for Type and Converted types of the lambda and method group. We should do the same but also verifying inferred nullability.
6. `var` and invocation sections of the spec are intentionally out of scope for this PR, is that correct?

Reply:
Thanks for the suggestions:
1. Added tests for `modopt` and `modreq`.
2. Added tests for missing `System.Action` and `System.Func`; added missing `System.Delegate` to the test plan.
3. Added test for parameter types and return type that are not valid type arguments.
4. No change.
5. Added nullability to test plan.
6. Correct, `var` and directly invoking a lambda expression are out of scope. Included in test plan.
7. A number of tests report the delegate type in the emitted executable.